### PR TITLE
Vin/mags layout

### PR DESCRIPTION
**Pull request by vinnie4k**
## Overview

This PR changes the layout of the “More Magazines” section as well as some UI and performance improvements.

### Shimmer Effect

- Created `ShimmerEffect` supporting view that could be used by applying `.shimmer` to a `View`
- Added a shimmer effect to articles, magazines, and publications skeleton when loading in

### Magazines List

- Optimized magazine loading by implementing pagination/infinite scrolling
- Changed the layout of the “More Magazines” section to a vertically scrolling grid with 2 columns
- Wrote a GraphQL query for fetching magazines by semester with limit and offset parameters

## Test Coverage

- Used different devices with varying screen sizes in Simulator to see if there were any UI issues
- In-depth playtesting to minimize any bugs that may be present

## Next Steps

Pagination for magazines in featured, publication view, and search results, but will wait for backend to support cover images before implementing.

## Related PRs or Issues

Magazines take a while to load. Will need backend to support cover image.

## Screenshots

<details>

  <summary>Shimmer Effect</summary>

  <table>
  <tr>
     <td>Before</td>
     <td>After</td>
  </tr>
  <tr>
    <td><video src="https://user-images.githubusercontent.com/75594943/227298938-c2425177-ace2-4417-8cf5-b67a2c5633f4.mp4" type="video/mp4" ></td>
<td><video src="https://user-images.githubusercontent.com/75594943/227299663-c71babae-47fa-43ce-ab4e-f931a839270d.mp4" type="video/mp4" ></td>
  </tr>
 </table>

</details>
<details>
  <summary>Magazines List</summary>
<table>
  <tr>
     <td>Before (this took forever to load)</td>
     <td>After (note that I was using Duffield wifi, usually loads within 3-5 seconds)</td>
  </tr>
  <tr>
    <td><video src="https://user-images.githubusercontent.com/75594943/227302460-7f63b352-35ad-4455-a23b-b34c6782310b.mp4" ></td>
<td><video src="https://user-images.githubusercontent.com/75594943/227302596-df3d64ce-b408-4974-a6f5-9b17e3e049d4.mp4" ></td>
  </tr>
 </table>
 

</details>

### CODE REVIEW

**Review by ReadeP — APPROVED** [review]
Looks good Vin! Wait for Han, Jennifer, or Vivian to also review before merging. Great work!

File: Volume/Supporting Views/SkeletonView.swift [hunk 1]
```diff
@@ -13,5 +13,6 @@ struct SkeletonView: View {
         Rectangle()
             .fill(Color.volume.veryLightGray)
             .transition(.opacity)
+            .shimmer(.init(tint: .gray.opacity(0.3), highlight: .white, blur: 30))
```
Comment (jjennifergu) [inline Volume/Supporting Views/SkeletonView.swift:16]: constants could be extracted into a Constants struct

File: Volume/View Models/MagazinesListViewModel.swift [hunk 2]
```diff
@@ -2,121 +2,220 @@
...
+        
+        // MARK: - Logic Constants
+        
         private struct Constants {
-            static let allMagazinesLimit: Double = 15
             static let featuredMagazinesLimit: Double = 7
-            static let animationDuration: Double = 0.1
+            static let semesterCountLimit: Double = 50  // TODO: Change this value when backend implements getMagazineSemesters
+            static let moreMagazinesLimit: Double = 4
+            static let allSemestersIdentifier: String = "all"
```
Comment (jjennifergu) [inline Volume/View Models/MagazinesListViewModel.swift:44]: suggesting to make semester identifier an enum for it to be more robust

File: Volume/Supporting Views/ShimmerEffect.swift [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+                                        .offset(x: size.width * moveTo)
+                                }
+                        }
+                        .mask {
+                            content
+                        }
+                    }
+                    .onAppear {
+                        DispatchQueue.main.async {
+                            moveTo = 1.5
```
Comment (jjennifergu) [inline Volume/Supporting Views/ShimmerEffect.swift:60]: you could move this into ShimmerConfig

File: Volume/View Models/MagazinesListViewModel.swift [hunk 4]
```diff
@@ -2,121 +2,220 @@
...
+        
+        private func loadMoreMagazines(with newMagazines: [Magazine]) {
+            switch moreMagazines {
+            case .none:
+                moreMagazines = newMagazines
+            case .some(let oldMagazines):
+                moreMagazines = oldMagazines + newMagazines
+            }
 
+            if newMagazines.count < Int(Constants.moreMagazinesLimit) {
```
Comment (jjennifergu) [inline Volume/View Models/MagazinesListViewModel.swift:210]: can just make moreMagazinesLimit of type Int when initialized

**Review by jjennifergu — APPROVED** [review]
this is really good vin! fixing all my bugs LOL😅 i just left some small suggestions but i'll approve the pr!

File: Volume/Supporting Views/ShimmerEffect.swift [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    @ViewBuilder
+    func shimmer(_ config: ShimmerConfig) -> some View {
+        self
+            .modifier(ShimmerEffectHelper(config: config))
+    }
+
+}
+
+struct ShimmerEffectHelper: ViewModifier {
```
Comment (vianxnguyen) [inline Volume/Supporting Views/ShimmerEffect.swift:21]: i'd just name this ShimmerEffect

File: Volume/Views/MainView/MagazinesList.swift [hunk 6]
```diff
@@ -2,38 +2,83 @@
...
+                moreMagazinesSection
+            }
+        }
+        .padding(.top)
+        .background(background)
+        .toolbar {
+            ToolbarItem(placement: ToolbarItemPlacement.navigationBarLeading) {
+                BubblePeriodText("Magazines")
+                    .font(.newYorkMedium(size: 28))
+                    .offset(y: 8)
```
Comment (vianxnguyen) [inline Volume/Views/MainView/MagazinesList.swift:48]: can factor out these constants

File: Volume/Views/MainView/MagazinesList.swift [hunk 7]
```diff
@@ -46,115 +91,90 @@ struct MagazinesList: View {
...
+                NavigationLink(Constants.navigationTitleKey, isActive: $viewModel.openMagazineFromDeeplink) {
+                    MagazineReaderView(initType: .fetchRequired(magazineID), navigationSource: .moreMagazines)
+                }
+                .hidden()
             }
         }
     }
     
 }
+
```
Comment (vianxnguyen) [inline Volume/Views/MainView/MagazinesList.swift:172]: can we delete this

File: Volume/View Models/MagazinesListViewModel.swift [hunk 8]
```diff
@@ -2,121 +2,220 @@
...
+                hasMoreMagazines = false
```
Comment (vianxnguyen) [inline Volume/View Models/MagazinesListViewModel.swift:211]: why is this set to false when number of new magazines is less than limit

File: Volume/Views/MainView/MagazinesList.swift [hunk 9]
```diff
@@ -46,115 +91,90 @@ struct MagazinesList: View {
...
-                            NavigationLink {
-                                MagazineReaderView(initType: .readyForDisplay(magazine), navigationSource: .moreMagazines)
-                            } label: {
-                                MagazineCell(magazine: magazine)
-                            }
+                    }
+                    
+                    if viewModel.hasMoreMagazines {
+                        ForEach(0..<2) { _ in
+                            MagazineCell.Skeleton()
```
Comment (vianxnguyen) [inline Volume/Views/MainView/MagazinesList.swift:116]: maybe factor out these magazine skeletons

**Review by vianxnguyen — APPROVED** [review]
this looks good!! 🎉

File: Volume/View Models/MagazinesListViewModel.swift (hunk 8 above)
Comment (vinnie4k) [reply Volume/View Models/MagazinesListViewModel.swift:211]: We want to fetch 4 magazines every time. If we ended up only fetching 3, this means that there are no more magazines to be fetched.

File: Volume/View Models/MagazinesListViewModel.swift (hunk 4 above)
Comment (vinnie4k) [reply Volume/View Models/MagazinesListViewModel.swift:210]: moreMagazinesLimit is used as an argument for a GraphQL query and cannot be an Int

**Pull request merged**